pkg/auth/providers/goauth/goauth_provider.go

```diff
@@ -0,0 +1,441 @@
+package goauth
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/rancher/norman/httperror"
+	"github.com/rancher/norman/types"
+	"github.com/rancher/norman/types/convert"
+	"github.com/rancher/rancher/pkg/auth/providers/common"
+	"github.com/rancher/rancher/pkg/auth/tokens"
+	corev1 "github.com/rancher/types/apis/core/v1"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/apis/management.cattle.io/v3public"
+	"github.com/rancher/types/client/management/v3"
+	publicclient "github.com/rancher/types/client/management/v3public"
+	"github.com/rancher/types/config"
+	"github.com/rancher/types/user"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/admin/directory/v1"
+	"google.golang.org/api/googleapi"
+	"google.golang.org/api/option"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"time"
+)
+
+const (
+	Name                 = "googleoauth"
+	userType             = "user"
+	groupType            = "group"
+	domainPublicViewType = "domain_public"
+)
+
+var scopes = []string{"openid", "profile", "email", admin.AdminDirectoryUserReadonlyScope, admin.AdminDirectoryGroupReadonlyScope}
+
+type goauthProvider struct {
+	authConfigs  v3.AuthConfigInterface
+	secrets      corev1.SecretInterface
+	goauthClient *GClient
+	userMGR      user.Manager
+	tokenMGR     *tokens.Manager
+	adminSvc     *admin.Service
+}
+
+func (g *goauthProvider) AuthenticateUser(ctx context.Context, input interface{}) (v3.Principal, []v3.Principal, string, error) {
+	login, ok := input.(*v3public.GoogleOauthLogin)
+	if !ok {
+		return v3.Principal{}, nil, "", fmt.Errorf("unexpected input type")
+	}
+	return g.loginUser(ctx, login, nil, false)
+}
+
+// loginUser takes as input the code; gets access_token and refresh_token in exhange; uses access_token to get user info
+// and groups (if allowed); and returns the user and group principals and oauth token
+func (g *goauthProvider) loginUser(c context.Context, googleOAuthCredential *v3public.GoogleOauthLogin, config *v3.GoogleOauthConfig, testAndEnableAction bool) (v3.Principal, []v3.Principal, string, error) {
+	var groupPrincipals []v3.Principal
```

Comment:
can we think of splitting this method into a several sub functions. It's a bit too long